File: D-DependencyInversion/Good_Switch.cpp
// ✅ GOOD — Follows Dependency Inversion Principle
// Both Switch (high-level) and LightBulb/Fan (low-level) depend on ISwitchable (abstraction).
// Switch doesn't care WHAT it controls — just that it implements ISwitchable.

#include <iostream>
#include <memory>

// The abstraction both sides depend on
class ISwitchable {
public:
    virtual void TurnOn()  = 0;
    virtual void TurnOff() = 0;
    virtual ~ISwitchable() = default;
};

// Low-level class 1
class LightBulb : public ISwitchable {
public:
    void TurnOn()  override { std::cout << "LightBulb is ON ✅\n"; }
    void TurnOff() override { std::cout << "LightBulb is OFF\n"; }
};

// Low-level class 2 — added WITHOUT touching Switch
class Fan : public ISwitchable {
public:
    void TurnOn()  override { std::cout << "Fan is ON ✅\n"; }
    void TurnOff() override { std::cout << "Fan is OFF\n"; }
};

// Low-level class 3 — still no changes to Switch needed
class SmartTV : public ISwitchable {
public:
    void TurnOn()  override { std::cout << "SmartTV is ON ✅\n"; }
    void TurnOff() override { std::cout << "SmartTV is OFF\n"; }
};

// High-level class depends ONLY on the abstraction ✅
class Switch {
    std::shared_ptr<ISwitchable> _device;

public:
    explicit Switch(std::shared_ptr<ISwitchable> device) : _device(device) {}

    void Operate(bool on) {
        if (on) _device->TurnOn();
        else    _device->TurnOff();
    }
};

// Usage:
// auto s1 = Switch(std::make_shared<LightBulb>());
// auto s2 = Switch(std::make_shared<Fan>());
// s1.Operate(true);   // LightBulb is ON ✅
// s2.Operate(false);  // Fan is OFF

File: D-DependencyInversion/Good_Switch.cs
// ✅ GOOD — Follows Dependency Inversion Principle
// Both Switch (high-level) and LightBulb/Fan (low-level) depend on ISwitchable (abstraction).
// Switch doesn't care WHAT it controls — just that it implements ISwitchable.

// The abstraction both sides depend on
public interface ISwitchable
{
    void TurnOn();
    void TurnOff();
}

// Low-level class 1
public class LightBulb : ISwitchable
{
    public void TurnOn()  => Console.WriteLine("LightBulb is ON ✅");
    public void TurnOff() => Console.WriteLine("LightBulb is OFF");
}

// Low-level class 2 — added WITHOUT touching Switch
public class Fan : ISwitchable
{
    public void TurnOn()  => Console.WriteLine("Fan is ON ✅");
    public void TurnOff() => Console.WriteLine("Fan is OFF");
}

// Low-level class 3 — still no changes to Switch needed
public class SmartTV : ISwitchable
{
    public void TurnOn()  => Console.WriteLine("SmartTV is ON ✅");
    public void TurnOff() => Console.WriteLine("SmartTV is OFF");
}

// High-level class — depends on abstraction, not concrete types
public class Switch
{
    private readonly ISwitchable _device;

    // ✅ Dependency is INJECTED — this is how ASP.NET Core DI works!
    public Switch(ISwitchable device)
        => _device = device;

    public void Operate(bool on)
    {
        if (on) _device.TurnOn();
        else    _device.TurnOff();
    }
}

// Usage — swap devices with zero changes to Switch:
// var lightSwitch = new Switch(new LightBulb());
// var fanSwitch   = new Switch(new Fan());
// var tvSwitch    = new Switch(new SmartTV());
//
// lightSwitch.Operate(true);  // LightBulb is ON ✅
// fanSwitch.Operate(true);    // Fan is ON ✅
// tvSwitch.Operate(false);    // SmartTV is OFF
//
// 💡 This is EXACTLY how ASP.NET Core registers services:
// builder.Services.AddScoped<ISwitchable, LightBulb>();
